LoRaEngine/LoraKeysManagerFacade/CreateEdgeDevice.cs

```diff
@@ -67,59 +71,89 @@ public CreateEdgeDevice(RegistryManager registryManager)
-                "'" + facadeKey + "'}");
-            var remoteTwin = await this.registryManager.GetTwinAsync(deviceName);
-
-            await this.registryManager.UpdateTwinAsync(deviceName, "LoRaWanNetworkSrvModule", twin, remoteTwin.ETag);
-
-            bool.TryParse(Environment.GetEnvironmentVariable("DEPLOY_DEVICE"), out bool deployEndDevice);
-
-            // This section will get deployed ONLY if the user selected the "deploy end device" options.
-            // Information in this if clause, is for demo purpose only and should not be used for productive workloads.
-            if (deployEndDevice)
+            catch (Exception)
             {
-                string otaaDeviceId = "47AAC86800430028";
-                var otaaDevice = new Device(otaaDeviceId);
-                await this.registryManager.AddDeviceAsync(otaaDevice);
-
-                var otaaEndTwin = new Twin();
-                otaaEndTwin.Properties.Desired = new TwinCollection(@"{AppEUI:'BE7A0000000014E2',AppKey:'8AFE71A145B253E49C3031AD068277A1',GatewayID:'',SensorDecoder:'DecoderValueSensor'}");
-                var otaaRemoteTwin = await this.registryManager.GetTwinAsync(otaaDeviceId);
-                await this.registryManager.UpdateTwinAsync(otaaDeviceId, otaaEndTwin, otaaRemoteTwin.ETag);
-
-                string abpDeviceId = "46AAC86800430028";
-                var abpDevice = new Device(abpDeviceId);
-                await this.registryManager.AddDeviceAsync(abpDevice);
-                var abpTwin = new Twin();
-                abpTwin.Properties.Desired = new TwinCollection(@"{AppSKey:'2B7E151628AED2A6ABF7158809CF4F3C',NwkSKey:'3B7E151628AED2A6ABF7158809CF4F3C',GatewayID:'',DevAddr:'0228B1B1',SensorDecoder:'DecoderValueSensor'}");
-                var abpRemoteTwin = await this.registryManager.GetTwinAsync(abpDeviceId);
-                await this.registryManager.UpdateTwinAsync(abpDeviceId, abpTwin, abpRemoteTwin.ETag);
+                // In case of an exception in device provisioning we want to make sure that we return a proper template if our devices are successfullycreated
+                var edgeGateway = await this.registryManager.GetDeviceAsync(deviceName);
+
+                if (edgeGateway == null)
+                {
+                    return PrepareResponse(HttpStatusCode.Conflict);
+                }
+
+                if (deployEndDevice)
+                {
+                    var abpDevice = await this.registryManager.GetDeviceAsync(AbpDeviceId);
+                    var otaaDevice = await this.registryManager.GetDeviceAsync(OtaaDeviceId);
+
+                    if (abpDevice == null || otaaDevice == null)
+                    {
+                        return PrepareResponse(HttpStatusCode.Conflict);
+                    }
+                }
+
+                return PrepareResponse(HttpStatusCode.OK);

```

Comment:
You can simplify by calling the PrepareResponse once (removing the call at line 143). Keep line 146.